### PR TITLE
Fix ZeRO-3 chunk synchronization in chunked losses

### LINKED ISSUE #6696: SFT chunked NLL stalls under ZeRO-3 with unequal rank chunk counts
## Summary

`SFTTrainer(loss_type="chunked_nll")` can stall in the first training step under DeepSpeed ZeRO-3 when assistant-only or completion-only masking leaves a different number of valid target tokens on each rank.

This is not a generic sparse-label or DDP failure. It occurs when valid-token counts cross different chunk boundaries, so ranks call the ZeRO-3 `lm_head` gather a different number of times.

## Environment

- TRL commit: `2396dfe5d2be7b18c0b615d80957d64ecdeb7cc0`
- PyTorch: 2.10.0
- Transformers: 5.13.1
- DeepSpeed: 0.18.8
- GPUs: 2

## Reproduction

Use `minimal_reproducer/repro_distributed_sparse_labels.py` with `zero3_bf16.json`.

The decisive case uses sequence length 2048, chunk size 256, and rank-local valid-token counts 64 and 320. Rank 0 executes one chunk while rank 1 executes two.

## Expected

The first step completes and each rank's loss uses only its non-`-100` labels.

## Actual

- NLL with the same data and ZeRO-3 completes; first step about 0.72 s.
- Chunked NLL without ZeRO-3 completes; first step about 0.84 s.
- ZeRO-3 chunked NLL with chunk counts 1/1 completes.
- ZeRO-3 chunked NLL with chunk counts 1/2 does not complete the first step within 120 s and exits by harness timeout 124.
- The same behavior reproduces with a Qwen3.5 VLM wrapper.

## Root Cause

`_chunked_cross_entropy_loss()` invokes `_chunk()` once per local valid-token chunk. `_chunk()` enters `maybe_gather_lm_head_ctx()`, which performs the ZeRO-3 gather for the sharded `lm_head`. Different local chunk counts therefore produce different collective call counts across ranks and stall distributed execution.

## Proposed Fix

Before the chunk loop, all-reduce the padded local chunk span with `MAX` under ZeRO-3. This gives every rank the same loop count. Extra local chunks contain only ignored labels and contribute zero loss. A fully masked rank must also remain in the loop when another rank has valid targets.

## Validation

- Failing LLM reproducer: timeout before patch, 0.73 s after patch.
- Failing Qwen3.5 VLM reproducer: over 5 min before patch, 1.83 s after patch.
- Fully masked rank versus two-chunk rank: pass after patch, masked-rank loss 0.
- Representative official regressions: 19 passed.

This is distinct from #3781, which concerned assistant masks being discarded with Liger in an older TRL version.

### PR DESCRIPTION
**Pull request by DimensionSTP**
# What does this PR do?

`SFTTrainer(loss_type="chunked_nll")` can stall under DeepSpeed ZeRO-3 when assistant-only or completion-only masking gives ranks different valid-token chunk counts. The same issue applies to the chunked divergence loss used by `DistillationTrainer`. Each local chunk gathers the sharded `lm_head`, so unequal loop counts desynchronize collectives.

This change synchronizes the padded chunk span with a rank-wise `MAX` in both chunked loss paths, keeps fully masked ranks in the shared loop, and safely clamps only the local fallback denominator. It adds regressions for unequal chunk counts and a fully masked rank in both trainers.

Validation: the minimal two-rank SFT reproducer times out before the fix and passes after it for Qwen3 LLM and Qwen3.5 VLM; the focused SFT and Distillation chunk-loss unit classes pass 41/41; Ruff format and diff checks pass.

Fixes #6696

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [ ] No AI usage: the PR was written entirely by a human.
- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag members/contributors who may be interested in your PR.

---

> [!NOTE]
> **Medium Risk**
> Touches distributed loss paths used by SFT and distillation training; incorrect sync logic could affect ZeRO-3 correctness or performance, but the change is narrow and covered by new regression tests.
> 
> **Overview**
> Fixes training hangs under **DeepSpeed ZeRO-3** when ranks see different numbers of valid tokens (e.g. completion-only / assistant-only masking), which previously led to mismatched chunk-loop counts and desynchronized `lm_head` gathers.
> 
> In **`_chunked_cross_entropy_loss`** (`SFTTrainer` / `chunked_nll`) and **`_chunked_divergence_loss`** (`DistillationTrainer`), the padded chunk span `n_padded` is now **`all_reduce`d with `MAX`** when ZeRO-3 is enabled and distributed is initialized, so every rank executes the same number of chunk iterations while local loss normalization still uses each rank’s valid-token count.
> 
> Adds unit tests that mock ZeRO-3 and `all_reduce` for unequal valid-token counts and for a fully masked rank, asserting expected `_chunk` invocations and safe backward.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 5d084f91755acecfc4f22fd4b40226f8fa932b4f. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/trainer/sft_trainer.py [hunk 1]
```diff
@@ -226,7 +231,7 @@ def _chunked_cross_entropy_loss(
         entropy_sum = entropy_sum + chunk_entropy
 
     if num_items_in_batch is None:
-        loss = loss / n_valid_tensor
+        loss = loss / n_valid_tensor.clamp_min(1)
```
[the remark this answers is not in the document]
Comment (DimensionSTP) [reply trl/trainer/sft_trainer.py]: Thanks for the review. _chunked_divergence_loss does not exist in either the PR base (2396dfe) or the current upstream main (24db2a8). The referenced lines 188–234 belong to _chunked_cross_entropy_loss, which this PR already updates. Therefore, no duplicated distillation chunk path remains to synchronize in the current codebase.

[the remark this answers is not in the document]
Comment (DimensionSTP) [reply trl/trainer/sft_trainer.py]: The original reply was accurate for the PR base at the time, where `_chunked_divergence_loss` did not yet exist. After rebasing onto current `main`, that duplicated chunked loss path is now present. Commit 91a5b51 applies the same ZeRO-3 rank-wise chunk-count synchronization, fully masked-rank participation, and safe local denominator handling to `DistillationTrainer`, with regressions for unequal valid-token counts and a fully masked rank. The focused SFT and Distillation chunk-loss tests pass 41/41.

**Comment by verma8076** [comment]
Heads up: this now conflicts with #6842, merged into main yesterday (2026-08-26). #6842 touches the exact same lines in both `sft_trainer.py::_chunked_cross_entropy_loss` and `distillation_trainer.py::_chunked_divergence_loss` — it removed the `if n_valid_tensor == 0:` early-return shortcut this PR's diff still has as context, and replaced the unconstrained `n_padded = (n_valid_tensor / chunk_size).ceil().to(torch.int64) * chunk_size` with `.clamp(min=1)` added before the cast, specifically so a fully-masked rank under context parallelism still runs one real chunk (needed for its zero loss to reach every trainable param during backward/grad-sync, rather than skip the loop entirely).

This PR's replacement `n_padded` computation (both copies) doesn't carry that `.clamp(min=1)`, and the `if n_padded == 0:` check it adds assumes the loop-skipping path #6842 removed. Rebasing naively risks resolving the conflict by keeping this PR's version and silently dropping the CP invariant #6842 just added — reintroducing a different collective-desync hang (this time under CP) while fixing the ZeRO-3 one.

The two fixes look straightforwardly composable, e.g.:

```python
n_padded = (n_valid_tensor / chunk_size).ceil().clamp(min=1).to(torch.int64) * chunk_size
if is_deepspeed_zero3_enabled() and torch.distributed.is_initialized():
    torch.distributed.all_reduce(n_padded, op=torch.distributed.ReduceOp.MAX)
```

with the `if n_padded == 0:` shortcut dropped entirely (since `clamp(min=1)` makes it unreachable, same as current main). Worth doing in both files given the repo's consistency-across-duplicated-trainers convention.

**Comment by DimensionSTP** [comment]
> Heads up: this now conflicts with #6842, merged into main yesterday (2026-08-26). #6842 touches the exact same lines in both `sft_trainer.py::_chunked_cross_entropy_loss` and `distillation_trainer.py::_chunked_divergence_loss` — it removed the `if n_valid_tensor == 0:` early-return shortcut this PR's diff still has as context, and replaced the unconstrained `n_padded = (n_valid_tensor / chunk_size).ceil().to(torch.int64) * chunk_size` with `.clamp(min=1)` added before the cast, specifically so a fully-masked rank under context parallelism still runs one real chunk (needed for its zero loss to reach every trainable param during backward/grad-sync, rather than skip the loop entirely).
> 
> This PR's replacement `n_padded` computation (both copies) doesn't carry that `.clamp(min=1)`, and the `if n_padded == 0:` check it adds assumes the loop-skipping path #6842 removed. Rebasing naively risks resolving the conflict by keeping this PR's version and silently dropping the CP invariant #6842 just added — reintroducing a different collective-desync hang (this time under CP) while fixing the ZeRO-3 one.
> 
> The two fixes look straightforwardly composable, e.g.:
> 
> ```python
> n_padded = (n_valid_tensor / chunk_size).ceil().clamp(min=1).to(torch.int64) * chunk_size
> if is_deepspeed_zero3_enabled() and torch.distributed.is_initialized():
>     torch.distributed.all_reduce(n_padded, op=torch.distributed.ReduceOp.MAX)
> ```
> 
> with the `if n_padded == 0:` shortcut dropped entirely (since `clamp(min=1)` makes it unreachable, same as current main). Worth doing in both files given the repo's consistency-across-duplicated-trainers convention.

Thanks for catching this. I rebased the PR on the current `main` and incorporated the invariant introduced in #6842.

The updated implementation now:

- keeps `.clamp(min=1)` so every rank executes at least one chunk;
- applies the ZeRO-3 `all_reduce(MAX)` after calculating the clamped padded span;
- removes the `n_padded == 0` early-return path;
- applies the same logic to both SFT chunked CE and Distillation chunked divergence.

Validation after the rebase:

- focused SFT and Distillation tests: 41/41 passed;
- 2-rank ZeRO-3 with 64/320 valid tokens: passed;
- 2-rank ZeRO-3 with a fully masked rank versus 320 valid tokens: passed, with masked-rank loss 0;
- Ruff import/format checks and `git diff --check`: passed.

Updated head: `8a4d4fc2660a5efd9de076c762f72b7e2737400e`

**Comment by DimensionSTP** [comment]
Hi @qgallouedec, following up on #6696: I rebased this PR onto the latest main (`94d6b7dd`) and resolved the Distillation import conflict. The updated head is `5d084f91`.

The fix preserves #6842’s `.clamp(min=1)` invariant and the recent chunked-loss changes. The ZeRO-3 issue remains distinct: ensuring at least one chunk does not equalize the number of chunk gathers across ranks. This PR applies `all_reduce(MAX)` to the clamped padded span in both SFT and Distillation.

Validation after the rebase:
- focused SFT and Distillation loss suites: 43 passed;
- two-rank SFT ZeRO-3: unequal chunk counts and a fully masked rank both completed three optimizer steps, with zero loss on the masked rank;
- Ruff lint/format, compilation, and `git diff --check`: passed.

Could you take another look when you have time? Thanks!

**Comment by DimensionSTP** [comment]
Hi @kashif, since you recently worked on chunked log-probabilities in #7065, would you be willing to review the distributed synchronization aspect of this fix?

The current revision includes the upstream changes and has passed focused CPU tests and two-rank ZeRO-3 SFT checks covering unequal chunk counts and a fully masked rank.

A review of the loop-bound and masking behavior would be especially helpful. Thank you!

**Comment by DimensionSTP** [comment]
Hi @qgallouedec and @kashif - this one is ready for another look.

The fix keeps every rank on the same chunk schedule under ZeRO-3, including ranks whose local target count would otherwise skip a collective. I have also carried the change over the chunked-loss updates from #6842, including its clamp invariant. CI is green.

The main point I would appreciate a second pair of eyes on is the collective placement in the SFT and Distillation paths.

**Comment by DimensionSTP** [comment]
Hi @behroozazarkhalili, your investigation of per-chunk gathers in #6172 looks relevant to one specific part of this fix. Would you be willing to review the collective schedule here?

This is a ZeRO-3 correctness issue, not the FSDP2 gather-count regression: ranks with different valid-token counts can enter different numbers of chunk iterations. The patch synchronizes the loop bound while preserving local masking, including a fully masked rank.

I checked main bb8af11c: the relevant loss bodies still have the rank-local bound, and this patch merges cleanly. Earlier validation includes two-rank ZeRO-3 runs; that latest-main check was static, not a new distributed run. A focused review of the synchronization placement would help move this PR toward maintainer approval.

**Comment by behroozazarkhalili** [comment]
I read the synchronization placement against main and ran the case your tests do not reach.

The collective is real and per chunk. On main `_chunk` (`sft_trainer.py:101`) enters `maybe_gather_lm_head_ctx(w, b)` for every chunk, which under ZeRO-3 with an untied head is `deepspeed.zero.GatheredParameters` (`trainer/utils.py:227`), and the non-reentrant checkpoint repeats that gather in the backward recompute. The chunk trip count is therefore a collective count, so a rank-local `n_padded` makes ranks disagree. Placing the `MAX` all-reduce right after `n_padded` and before the loop is the only correct spot: it needs the local count and must precede the first gather, and `MAX` (rather than a mean or the local value) guarantees no rank leaves the loop early. The distillation loop has the same per-chunk gather on the teacher head, so the second site is warranted.

The case the tests leave out is a rank whose flattened token count is shorter than the peer's `n_padded`, where the extra iterations slice empty tensors. I ran that on your head (5d084f91, CPU, mocked `all_reduce` forcing 16 on a rank with 8 positions and 4 valid targets): 4 chunk calls with token counts [4, 4, 0, 0], loss finite and bit-identical to the unsynchronized value (5.723011), finite gradients on both `hidden` and the weight. So the empty chunks add exactly zero and still issue their gathers.

The ZeRO-3-only guard matches where the collective exists. Under FSDP2 `maybe_gather_lm_head_ctx` is a null context, and #6172 measured on two H100s that the chunk loop issues no per-chunk all-gather there, so unequal chunk counts do not desynchronize that path.

What I did not verify: a two-rank ZeRO-3 run of this head (your timeout-before, pass-after result stands as your measurement, not mine), and CI still never exercises the collective because both tests mock `all_reduce`; a `zero3` case in `tests/distributed` would close that gap if the maintainers want it covered.
